### PR TITLE
[YUNIKORN-941] split scheduler and admission controller deployment

### PR DESCRIPTION
**Pull request by kingamarton**

### CODE REVIEW

**Comment by wilfred-s** [comment]
Changes look good. How has this been tested?
Is there any impact on the release build code that we use to generate the helm chart?

File: helm-charts/yunikorn/templates/admission.yaml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            - name: webhook-tls-certs
+              mountPath: /run/secrets/tls
+              readOnly: true
+          env:
+            - name: NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: POLICY_GROUP
+              value:  {{ .Values.admission_controller_image.policyGroup }}
```
Comment (yuchaoran2011) [inline helm-charts/yunikorn/templates/admission.yaml:77]:
```suggestion
              value:  {{ .Values.admissionControllerImage.policyGroup }}
```
Minor nit: can we take this opportunity to change the name of this variable to use camelcase, conforming to the Helm [best practice](https://helm.sh/docs/chart_best_practices/values/#naming-conventions)?

**Review by yuchaoran2011 — CHANGES_REQUESTED** [review]

File: helm-charts/yunikorn/templates/admission.yaml (hunk 1 above)
Comment (wilfred-s) [reply helm-charts/yunikorn/templates/admission.yaml:77]: That sounds like a real good idea

**Comment by kingamarton** [comment]
> Changes look good. How has this been tested? Is there any impact on the release build code that we use to generate the helm chart?

@wilfred-s I tested it manually: built the new images, installed the new charts and tried to change the configmap and checked if the validation hook is working properly, then submitted some apps without any scheuler name and checked if te mutation hook was working and the scheduler name was populated.

No, it should not have any effect on the release build code, since the admission controller will be in the generated package. 

However this is working now, I would go forward and try to moove the admission controller part into sepparate helm chart, that way it will be completely independent from the rest of the YK deployment. Now, with this change we can update the scheduler without updating the adm. controller just by using kubectl commands. If we update it using helm commands, the admission controller will be updated as well, since it is in the same helm deployment.

**Review by yuchaoran2011 — COMMENTED** [review]
I'm not sure if placing the admission controller in its own Helm chart is a good idea since it will require two Helm commands to install YK. Having both the scheduler and the admission controller in a single chart enables one-command install and still allows each component to be updated without touching the other

**Comment by wilfred-s** [comment]
As long as we can upgrade the admission controller first without issues I am ok with keeping it in one chart. The reason why we need to be able to upgrade separately is that we do not want to get into the situation that some pods get scheduled by the default scheduler because the admission controller is not there and the scheduler is gone too.

**Comment by kingamarton** [comment]
I checked today how Helm install/update works. It has a hardcoded order of installing different kinds and manifests, but I haven't seen anything about an order of the same type, however we can force an order in two ways:
- by using annotations and set pre-install hook, so we make sure that the admission controller is installed first
- by using init containers with a simple sleep command. Set a higher value for the scheduler.  See the details in the following article: https://stackoverflow.com/questions/59653816/helm-install-in-specific-order-for-deployment

To be honest, none of this solutions seems very reliable. I would rather install them manually, so I can make sure that the admission controller is up and running before the scheduler.
@yuchaoran2011  do you have any ideas on how this order can be forced without splitting the chrts?

**Comment by yuchaoran2011** [comment]
@kingamarton Thanks for looking into the options. Neither the pre-install hook nor init containers seem like an elegant solution. I agree with what @wilfred-s said on the need to keep the admission controller running while the scheduler is being upgraded, but I think it's still preferable to keep YuniKorn installation as a one-command procedure. Having two standalone Helm charts will be inconvenient for users. 

Now that you have refactored the admission controller into its own manifest files, is it possible to make it a [sub-chart](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/) of the overall YuniKorn chart? That way, we'll have a single chart that's effectively implemented using two charts: The scheduler parent chart and the admission controller sub-chart. What do you think?

**Comment by kingamarton** [comment]
> @kingamarton Thanks for looking into the options. Neither the pre-install hook nor init containers seem like an elegant solution. I agree with what @wilfred-s said on the need to keep the admission controller running while the scheduler is being upgraded, but I think it's still preferable to keep YuniKorn installation as a one-command procedure. Having two standalone Helm charts will be inconvenient for users.
> 
> Now that you have refactored the admission controller into its own manifest files, is it possible to make it a [sub-chart](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/) of the overall YuniKorn chart? That way, we'll have a single chart that's effectively implemented using two charts: The scheduler parent chart and the admission controller sub-chart. What do you think?

Thank you @yuchaoran2011! The subcharts sound exactly what we need. I will modify the code and do some tests to make sure that it will solve our issue.

**Review by yuchaoran2011 — APPROVED** [review]
LGTM. Thanks for working on it! Did you already test it? If so, it's all good to be merged

**Comment by kingamarton** [comment]
> LGTM. Thanks for working on it! Did you already test it? If so, it's all good to be merged

Yes @yuchaoran2011 , I tested it and left comment in the Jira with the necessary steps in case of an upgrade. Thank you for the review, I am going to merge it. Can you please check the shim side changes as well?

**Pull request merged**